### PR TITLE
Made EventDispatcher optional.

### PR DESCRIPTION
**Pull request by darkdef**
For simple usage, if you don't need event dispatcher for translation events.

| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌
| Fixed issues  |

### CODE REVIEW

File: src/Translator.php [hunk 1]
```diff
@@ -26,14 +26,14 @@ class Translator implements TranslatorInterface
...
+     * @param EventDispatcherInterface $eventDispatcher Event dispatcher for translation events. Null for none.
      * @param string|null $fallbackLocale Locale to use if message for the locale specified was not found. Null for none.
      */
     public function __construct(
         Category $defaultCategory,
         string $locale,
-        EventDispatcherInterface $eventDispatcher,
-        string $fallbackLocale = null
+        ?EventDispatcherInterface $eventDispatcher = null,
+        ?string $fallbackLocale = null
```
Comment (samdark) [inline src/Translator.php:36]: What's the most common use case?

1. Using fallback locale without events.
2. Using events without fallback locale.

Comment (darkdef) [reply src/Translator.php:36]: May be switch params? Or adding method for EventDispatcher - withEventDispatcher (without immutable)

Comment (samdark) [reply src/Translator.php:36]: Yes. I suggest swapping params.

File: src/Translator.php [hunk 2]
```diff
@@ -80,15 +80,19 @@ public function translate(
...
             return $id;
         }
 
         $sourceCategory = $this->categories[$category];
         $message = $sourceCategory->getMessage($id, $locale, $parameters);
 
         if ($message === null) {
-            $this->eventDispatcher->dispatch(new MissingTranslationEvent($sourceCategory->getName(), $locale, $id));
+            if ($this->eventDispatcher !== null) {
+                $this->eventDispatcher->dispatch(new MissingTranslationEvent($sourceCategory->getName(), $locale, $id));
```
Comment (samdark) [inline src/Translator.php:94]: Seems our tests for events aren't complete. See Mutant below.

**Comment by samdark** [comment]
Overall direction is alright. Let's finish this one.

File: tests/EventTest.php [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+declare(strict_types=1);
+
+use PHPUnit\Framework\TestCase;
+use Yiisoft\Translator\Event\MissingTranslationEvent;
+
+class EventTest extends TestCase
+{
+    public function testMissingTranslationEvent(): void
+    {
+        $event = new MissingTranslationEvent('app', 'en', 'missed_message');
```
Comment (samdark) [inline tests/EventTest.php:12]:
```suggestion
        $event = new MissingTranslationEvent('app', 'en', 'missing_message');
```

File: tests/EventTest.php [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        $this->assertEquals('app', $event->getCategory());
+        $this->assertEquals('en', $event->getLanguage());
+        $this->assertEquals('missed_message', $event->getMessage());
```
Comment (samdark) [inline tests/EventTest.php:15]:
```suggestion
        $this->assertEquals('missing_message', $event->getMessage());
```

File: tests/TranslatorTest.php [hunk 5]
```diff
@@ -188,12 +238,32 @@ public function testTranslationMissingCategory()
...
         $translator->translate('miss', [], 'miss');
     }
 
+    public function testTranslationMissingMessage(): void
+    {
+        $eventDispatcher = $this->getMockBuilder(EventDispatcherInterface::class)->getMock();
+        $eventDispatcher
+            ->expects($this->once())
+            ->method('dispatch')
+            ->with(new MissingTranslationEvent('app', 'en', 'missed_message'));
```
Comment (samdark) [inline tests/TranslatorTest.php:254]:
```suggestion
            ->with(new MissingTranslationEvent('app', 'en', 'missing_message'));
```

File: tests/TranslatorTest.php [hunk 6]
```diff
@@ -188,12 +238,32 @@ public function testTranslationMissingCategory()
...
+
+        /** @var EventDispatcherInterface $eventDispatcher */
+        $translator = new Translator(
+            $this->createCategory('app', $this->getMessages()),
+            'en',
+            null,
+            $eventDispatcher
+        );
+
+        $translator->translate('missed_message', [], 'app');
```
Comment (samdark) [inline tests/TranslatorTest.php:264]:
```suggestion
        $translator->translate('missing_message', [], 'app');
```

**Review by samdark — APPROVED** [review]
Looks good 👍 I wonder if you really meant "missed" :)

**Comment by samdark** [comment]
Thanks!

**Pull request merged**